src/urlpattern.cpp

```diff
@@ -0,0 +1,201 @@
+#include "ada/unicode.h"
+#include "ada/urlpattern.h"
+#include <forward_list>
+
+namespace ada {
+
+ada_really_inline bool is_valid_name_code_point(const char32_t& c,
+                                                bool is_first) noexcept {
+  return is_first ? unicode::is_valid_identifier_start(c)
+                  : ada::unicode::is_valid_identifier_part(c);
+}
+
+ada_really_inline bool is_ascii(char32_t c) { return c < 0x80; }
+
+enum class TOKEN_TYPE {
+  OPEN,
+  CLOSE,
+  REGEXP,
+  NAME,
+  CHAR,
+  ESCAPED_CHAR,
+  OTHER_MODIFIER,
+  ASTERISK,
+  END,
+  INVALID_CHAR
+};
+
+enum class POLICY { STRICT, LENIENT };
+
+struct TOKEN {
+  TOKEN_TYPE token;
+  size_t start;
+  size_t end;
+};
+
+ada_really_inline std::forward_list<TOKEN> tokenize(std::string_view input,
+                                                    POLICY policy) {
+  // Need to deal with Unicode points, so convert to std::u32string_view
+  std::forward_list<TOKEN> tokens{};
+  size_t input_size = input.size();
+
+  // TODO: convert input to utf32
+  const auto error_or_invalid = [&](const std::string_view msg, size_t start,
+                                    size_t end) {
+    if (policy != POLICY::LENIENT) {
+      throw std::invalid_argument(std::string(msg));
+    }
+    tokens.push_front({TOKEN_TYPE::INVALID_CHAR, start, end});
+  };
+
+  size_t index = 0;
+  while (index < input_size) {
```

Comment:
Can we iterate through a while loop and find the next available item through something like `buffer.find_first_of("*+?\\{}:")`?

Reply:
Great idea!